config/store.go

```diff
@@ -188,124 +196,155 @@ func (s *Store) Set(newCfg *model.Config) (*model.Config, error) {
 		newCfg.FeatureFlags = nil
+		newCfgNoEnv.FeatureFlags = nil
 	}
 
-	if err := s.backingStore.Set(newCfg); err != nil {
+	if err := s.backingStore.Set(newCfgNoEnv); err != nil {
 		return nil, errors.Wrap(err, "failed to persist")
 	}
 
-	if err := s.loadLockedWithOld(oldCfg, &unlockOnce); err != nil {
-		return nil, errors.Wrap(err, "failed to load on save")
+	newCfg = applyEnvironmentMap(newCfgNoEnv, GetEnvironment())
+	fixConfig(newCfg)
+	if err := newCfg.IsValid(); err != nil {
+		return nil, errors.Wrap(err, "new configuration is invalid")
+	}
+
+	hasChanged, err := confsDiff(oldCfg, newCfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to compare configs")
+	}
+
+	// We restore the previously cleared feature flags sections back.
+	if s.readOnlyFF {
+		oldCfg.FeatureFlags = oldCfgFF
+		newCfg.FeatureFlags = oldCfgFF
+		newCfgNoEnv.FeatureFlags = oldCfgNoEnvFF
+	}
+
+	s.configNoEnv = newCfgNoEnv
+	s.config = newCfg
+
+	unlockOnce.Do(s.configLock.Unlock)
+
+	if hasChanged {
+		s.invokeConfigListeners(oldCfg, newCfg.Clone())
 	}
 
 	return oldCfg, nil
 }
 
-func (s *Store) loadLockedWithOld(oldCfg *model.Config, unlockOnce *sync.Once) error {
+// Load updates the current configuration from the backing store, possibly initializing.
+func (s *Store) Load() error {
+	s.configLock.Lock()
+	var unlockOnce sync.Once
+	defer unlockOnce.Do(s.configLock.Unlock)
+
+	oldCfg := &model.Config{}
+	if s.config != nil {
+		oldCfg = s.config
+	}
+
 	configBytes, err := s.backingStore.Load()
 	if err != nil {
 		return err
 	}
 
-	loadedConfig := &model.Config{}
+	loadedCfg := &model.Config{}
 	if len(configBytes) != 0 {
-		if err = json.Unmarshal(configBytes, &loadedConfig); err != nil {
+		if err = json.Unmarshal(configBytes, &loadedCfg); err != nil {
 			return jsonutils.HumanizeJSONError(err, configBytes)
 		}
 	}
 
-	loadedFeatureFlags := loadedConfig.FeatureFlags
-
 	// If we have custom defaults set, the initial config is merged on
 	// top of them and we delete them not to be used again in the
 	// configuration reloads
 	if s.configCustomDefaults != nil {
 		var mErr error
-		loadedConfig, mErr = Merge(s.configCustomDefaults, loadedConfig, nil)
+		loadedCfg, mErr = Merge(s.configCustomDefaults, loadedCfg, nil)
 		if mErr != nil {
 			return errors.Wrap(mErr, "failed to merge custom config defaults")
 		}
 		s.configCustomDefaults = nil
 	}
 
-	loadedConfig.SetDefaults()
-
-	s.configNoEnv = loadedConfig.Clone()
-	fixConfig(s.configNoEnv)
+	// We set the SiteURL to empty (if nil) so that the following call to
+	// SetDefaults() will generate missing data. This avoids an additional write
+	// to the backing store.
+	if loadedCfg.ServiceSettings.SiteURL == nil {
+		loadedCfg.ServiceSettings.SiteURL = model.NewString("")
+	}
 
-	loadedConfig = applyEnvironmentMap(loadedConfig, GetEnvironment())
+	// Setting defaults allows us to accept partial config objects.
+	loadedCfg.SetDefaults()
 
-	fixConfig(loadedConfig)
+	loadedCfgNoEnv := loadedCfg
```

Comment:
I was wondering why is it not cloned here since it's just a pointer, but then realized that `applyEnvironmentMap` does the cloning internally. Wondering if it might be better to clone it explicitly and let `applyEnvironmentMap` just work on the pointer? 0/5.

Alternatively, just add a comment mentioning what's happening here so that it's clear for the reader.

Reply:
Yeah. At the beginning of this PR I've attempted to remove all the internal clones but it was getting harder to keep track of what was safe and what wasn't so eventually I left it there. Make sense to add a comment.